src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -252,6 +324,83 @@ public void ConnectHeaders_SupplyHeaders_ToDownloadArraySizeSameAsNumberOfHeader
             Assert.Equal(headersToDownloadCount, peer2Headers.Count);
         }
 
+        /// <summary>
+        /// Issue 8 @ Create chained header tree component #1321
+        /// We have a chain and peer B is on the best chain.
+        /// After that peer C presents an alternative chain with a forkpoint below B's Tip.
+        /// After that peer A prolongs C's chain (but sends all headers from the fork point) with a few valid and one invalid.
+        /// Make sure that C's chain is not removed - only headers after that.
+        /// </summary>
+        [Fact]
+        public void ConnectHeaders_MultiplePeersWithForks_CorrectTip()
+        {
+            var testContext = new TestContext();
+            ChainedHeaderTree chainedHeaderTree = testContext.CreateChainedHeaderTree();
+            ChainedHeader chainTip = testContext.ExtendAChain(10);
+            chainedHeaderTree.Initialize(chainTip, true);
+
+            List<BlockHeader> listOfExistingHeaders = testContext.ChainedHeaderToList(chainTip, 10);
+            chainedHeaderTree.ConnectNewHeaders(2, listOfExistingHeaders);
+            ChainedHeader peerTwoTip = chainedHeaderTree.GetChainedHeaderByPeerId(2);
+
+            // Peer B / Peer 2 is on the best chain.
+            peerTwoTip.HashBlock.Should().BeEquivalentTo(chainTip.HashBlock);
+            peerTwoTip.ChainWork.Should().BeEquivalentTo(chainTip.ChainWork);
+
+            // Peer C / Peer 3 presents an alternative chain with a forkpoint below B's Tip.
+            BlockHeader forkedBlock = testContext.GetNewBlock(chainTip.Previous.Previous).Header;
+            chainedHeaderTree.ConnectNewHeaders(3, new List<BlockHeader>() { forkedBlock });
+            ChainedHeader peerThreeTip = chainedHeaderTree.GetChainedHeaderByPeerId(3);
+
+            ChainedHeader fork = chainTip.FindFork(peerThreeTip);
+            fork.Height.Should().BeLessThan(peerTwoTip.Height);
+            
+            // Peer A / PEER_ONE prolongs C's chain.
+            const int numberOfBlocksToExtend = 7;
+            ChainedHeader peerOneTip = testContext.ExtendAChain(numberOfBlocksToExtend, peerThreeTip);
+            List < BlockHeader > listOfPeerOnesHeaders = testContext.ChainedHeaderToList(peerOneTip, numberOfBlocksToExtend + 2 /* (forked 2 back) */);
+            
+            // Sends all headers from the fork point
+            fork.HashBlock.Should().BeEquivalentTo(listOfPeerOnesHeaders[0].GetHash());
+
+            // With few valid and one invalid.
+            foreach (var header in listOfPeerOnesHeaders.GetRange(3, 6))
+            {
+                testContext.ChainStateMock.Setup(x => x.MarkBlockInvalid(header.GetHash(), null));
+            }
```

Comment:
1) only for the last block in that list
2) you need to mock this.chainedHeaderValidator.ValidateHeader to throw exception for that particular block. ChainStateMock would not work here

Reply:
Done